### ISSUE TITLE
Split location field into state and city (at least)

### ISSUE THREAD

**Issue by davidskalinder** [davidskalinder/mpeds-coder#14]
Should be straightforward in UI, but needs to be carefully integrated into pass 2...

**Comment by davidskalinder** [comment]
Fixed in branch split_location, needs testing and final merge

**Comment by davidskalinder** [comment]
@olderwoman, @matthewsmorgan, @limchaeyoon: This should now be fixed and available in the testing implementation, so if y'all could check it out and comment here if it looks okay, then I'll get this feature merged into the production version.

**Comment by davidskalinder** [comment]
@olderwoman sez: Oh, So I was never the one who thought separating the state and city fields was important, and find it sort of annoying to have to keep typing out state names, but it worked ok. We are still subject to typographical errors.

I wrote in “all 50 states” and “uncertain” as the state in some cases.

**Comment by davidskalinder** [comment]
> @olderwoman sez: Oh, So I was never the one who thought separating the state and city fields was important, and find it sort of annoying to have to keep typing out state names, but it worked ok. We are still subject to typographical errors.
> 
> I wrote in “all 50 states” and “uncertain” as the state in some cases.

So I think the big advantage of splitting them out is that states should be much easier to decide and to reconcile than cities (or any other finer-grained location).  Furthermore, I'm guessing that the number of protests at a given time in a given state will be small, so being able to quickly organize events by state + date should dramatically ease tasks like identifying which events might be "the same".  Also obviously it was an easy change to make, so I figured the value was high for a low cost.

Having said that, it's interesting that even in this little sample dataset there was stuff that was either clearly multi-state or not easily decidable.  I guess the uncertain ones can be handled as part of issue #11?  But multi-state is a little trickier.  I guess it raises the question of whether something that occurs in all 50 states is one "event"?  Or 50!?

As for typos (or, for that matter whole words, abbreviations, and even the dreaded "Washington" ambiguity), I think the options are:

1. Put in a dropdown list
2. Keep it open in the (probably reasonable) hope that it usually won't be too hard to clean

Option 1 might not be too bad?  I'll check it out.  But in the case that it's difficult then I think option 2 wouldn't be a disaster either?

So I think the outstanding question for @olderwoman (and possibly @matthewsmorgan and @limchaeyoon as well) is how we want to count things that happen in multiple states.  I guess I'm a little hesitant to call it all one "event" since each thing in each place would be so different from one another, but I'm curious what y'all think...

**Comment by olderwoman** [comment]
I don't object to splitting state & city. Not true that it is always easier to identify state than city. "Uncertain" state was because I was coding what was obviously a press release by the NAACP and it really was not clear where they were when they released it as both national and Wisconsin NAACP officials were quoted. 

As far as multiple locations, is is VERY common for an article to say "there were protests in all 50 states" or "there were  protests in more than 20 cities" but not to list them individually. 

If you have been reading my methods articles (ahem), you will recall that I have actually explicitly discussed the matter of how I have been coding these ambiguous aggregate events and why I've been doing it. This is not a new issue. I have been explicitly arguing for the need to recognize these aggregate events that don't have explicit locations.

**Comment by davidskalinder** [comment]
> I don't object to splitting state & city. Not true that it is always easier to identify state than city. 

Well, even if it often is it seems like it'll help to more quickly identify those cases?

> "Uncertain" state was because I was coding what was obviously a press release by the NAACP and it really was not clear where they were when they released it as both national and Wisconsin NAACP officials were quoted.
> 
> As far as multiple locations, is is VERY common for an article to say "there were protests in all 50 states" or "there were protests in more than 20 cities" but not to list them individually.
> 
> If you have been reading my methods articles (ahem), you will recall that I have actually explicitly discussed the matter of how I have been coding these ambiguous aggregate events and why I've been doing it. This is not a new issue. I have been explicitly arguing for the need to recognize these aggregate events that don't have explicit locations.

Haha, indeed.  (I haven't been following all the methods discussions closely, and if I did in fact read that in one of them, I'd obviously forgotten!)  I suppose the old field that asked for "Location (City, State/Province, Country)" would have had the same problem, huh...  Could you perhaps send me the methods paper that's the most prescriptive on this point?  I could speculate that we might want something like different ways of handling reports of events with definite locations vs. reports of events with indefinite locations, but you're right that I should read the stuff you've written about this more carefully since I'm now at the stage where I can use it to make changes...

**Comment by olderwoman** [comment]
The latest version of the methods paper is in the google drive/writing/methods paper. I also just discovered that I uploaded the wrong slide deck, which I'm about to fix. It has a few things that were not in the version I uploaded. 

Re city vs state, sometimes a newspaper will just give the city and not the state, so the coder has to actually know that Chicago is in Illinois and Newark is in New Jersey.

**Comment by davidskalinder** [comment]
> The latest version of the methods paper is in the google drive/writing/methods paper. I also just discovered that I uploaded the wrong slide deck, which I'm about to fix. It has a few things that were not in the version I uploaded.

Got it, thanks.  I probably won't get to it until tomorrow, but I'll have a look.

> 
> Re city vs state, sometimes a newspaper will just give the city and not the state, so the coder has to actually know that Chicago is in Illinois and Newark is in New Jersey.

Yeah, not to mention Springfield.  I suppose the hope is that there will be a lot fewer ways for a "state" field to go wrong than for a "state and city" field to go wrong, and therefore fewer hard cases to clean up just to get to the state level.  But we'll never get everything.

**Comment by matthewsmorgan** [comment]
Just tried out this change. Adding to Pam's comments, I think whatever we end up with it would be nice to have more explicit instructions for the Pass 1 coder in the actual interface. So, if we stick with separate City and State, we could have in the interface: State (Acronym or full name; if multiple, type "Multiple")

These instructions can also of course be in the codebook, but it's nice to take away the extra thinking involved in trying to remember, "Do I need to type the full state name? Is an acronym okay?" And then if you're a diligent pass 1 coder going back to the codebook to check. Let's take away that extra work!

**Comment by davidskalinder** [comment]
So, increasingly I'm thinking that a drop-down will be the way to go with this.  I'm thinking that it should look like this:

State/Territory
1. No identifiable state/territory
2. Multiple identifiable states/territories
3. Somebody's standardized and alphabetized list of US states and territories

Then a text box for City and then another text box for "Other location" (to get things like "Adams St. Bridge" but also, e.g., states that for some reason don't fit into the list above even though my hope is that it's logically exhaustive).

This would have the advantages of standardizing the state and territory data; of making clear what the pass 1 coders should enter; of prompting them for things like Washington state vs DC; of providing them with a list of US territories; and of speed, since you can navigate an alphabetized drop-down by pressing the first letter of the word.

@olderwoman, I think items 1 and 2 on the list above also handle the cases you raise above and in the methods paper?  But would be good to hear people's thoughts on whether this setup would work or whether I'm missing something.

**Comment by olderwoman** [comment]
I like this. Technically in Census Speak City would be Place (i.e. villages, towns, counties, townships are also places and get mentioned in stories about less populous areas) but I doubt that would confuse anybody at all.

**Comment by davidskalinder** [comment]
Okay great, I'll proceed with that unless I hear otherwise.  We could call it "place" in the data and "City" to the coders (in the UI) if we want?  Though those labels will be very easy to change (until we have real data collected under the old label at the very least).  Ultimately I think we should use whatever data labels are clearest for us...

Also @olderwoman, you don't happen to know of a good list of states and territories somewhere, do you?  If not then I'll get Googling and you can review whatever I find...

**Comment by olderwoman** [comment]
John actually downloaded the Census FIPS files which includes such a list. He is traveling today until later tonight but could doubtless share it with you. Also if you google FIPS states there are about 100 copies on the Internet. There are also FIPS place and county codes. A good list will include Puerto Rico, Guam, etc.

**Comment by davidskalinder** [comment]
Okay that's what I'll look for then.  Thanks!

**Comment by davidskalinder** [comment]
Hey @alexhanna, I'm looking through the js to put in a select control for this and wanted to check something vaguely related to this issue: do I have it right that the "Save" button doesn't do anything?  Or, to be more precise, it doesn't save anything -- just gets rid of all the event controls?  It looks like everything just gets ajax'd into the database on change/blur/etc and so the save button is a nice placebo?

**Comment by davidskalinder** [comment]
Okay, looks like this will be trickier than I thought since it looks like we'll need new handlers for select controls.  So, some notes for myself that everybody else can ignore except for @alexhanna if she wants to not ignore them:

Will need to put a new listener in like this: https://github.com/davidskalinder/mpeds-coder/blob/570b5bc098ef9e1ec81b4f04f1a1237033b01643/static/ec.js#L60  Not sure if the block should be specified or not (it's currently specified for the radios, but not for the checkboxen).  Otherwise I assume the radio handlers are the best model since radios are the most like select controls.

Then we also need a handler like this: https://github.com/davidskalinder/mpeds-coder/blob/570b5bc098ef9e1ec81b4f04f1a1237033b01643/static/shared.js#L256  Or maaaaybe we could just cheekily use the same handler as for radios since they're so much alike?  Should try that first and see if it works.  If so then I'm inclined to use the same handler for both types to avoid a stale duplicate.  Ideally, it should be renamed so it's clear that it's for both types, but that's a nice-to-have.

**Comment by alexhanna** [comment]
> Hey @alexhanna, I'm looking through the js to put in a select control for this and wanted to check something vaguely related to this issue: do I have it right that the "Save" button doesn't do anything? Or, to be more precise, it doesn't save anything -- just gets rid of all the event controls? It looks like everything just gets ajax'd into the database on change/blur/etc and so the save button is a nice placebo?

That is correct! Save doesn't actually do anything other than get rid of the event block. Everything is already saved as soon as the coder enters it

> Or maaaaybe we could just cheekily use the same handler as for radios since they're so much alike? Should try that first and see if it works. If so then I'm inclined to use the same handler for both types to avoid a stale duplicate. Ideally, it should be renamed so it's clear that it's for both types, but that's a nice-to-have.

You might be able to get away with using the same function for the radio handler because all this really does is ensure that you're deleting all prior instances of the variable and inserting a new one. That holds true for location.

**Comment by davidskalinder** [comment]
> Not sure if the block should be specified or not (it's currently specified for the radios, but not for the checkboxen).

Answer to self: probably not.  It's specified for radios because in the presets tab, choosing a radio option expands a new set of checkboxen.

**Comment by davidskalinder** [comment]
> > Or maaaaybe we could just cheekily use the same handler as for radios since they're so much alike? Should try that first and see if it works. If so then I'm inclined to use the same handler for both types to avoid a stale duplicate. Ideally, it should be renamed so it's clear that it's for both types, but that's a nice-to-have.
> 
> You might be able to get away with using the same function for the radio handler because all this really does is ensure that you're deleting all prior instances of the variable and inserting a new one. That holds true for location.

Okay @alexhanna, another question for you about this: I notice that the selectRadio and the storeText functions are identical, right down to some mentions of radios in storeText's strings that should probably be changed.  That makes me think that when you wrote this you decided that it'd be better to have different handlers for each control type even if they're duplicated, rather than have multiple control types share the same handler?  So maybe I should just copy it again, call it selectOption or some such, and write a new listener line to call that one, rather than simply writing a listener to call selectRadio?  Or should we go the avoid-duplication route and share whenever possible (and maybe someday combine them into a single changeVariable handler or some such)?

Also, do you want this in upstream (the select/option handlers that is, not the changes to event-creator-block)?

**Comment by davidskalinder** [comment]
> > Not sure if the block should be specified or not (it's currently specified for the radios, but not for the checkboxen).
> 
> Answer to self: probably not. It's specified for radios because in the presets tab, choosing a radio option expands a new set of checkboxen.

Rejoinder to self: though OTOH, leaving it unspecified might set it globally and do something weird to the admin page or something.  Best to be conservative and specify only the sections where it's needed I think.  Listeners are cheap.

**Comment by davidskalinder** [comment]
This should now be fixed, with the list of states/territories loaded from a yaml file, except that the order of the states is stupidly random.  So this now depends on #15.

**Comment by davidskalinder** [comment]
If this doesn't want to wait on #15,  we could also inelegantly hard-code the list of states as an OrderedDict...

**Comment by davidskalinder** [comment]
Should now be fixed in branch split_location.

**Comment by davidskalinder** [comment]
@olderwoman and @matthewsmorgan, this issue should now be fixed and up and running in the testing implementation if you want to have a look?  It's been a while since we've tested anything, so if you need the logins they're in the Technical documentation > MAI folder; let me know if you need any other info.

I've checked that the data is being recorded correctly in the database, but I realize there might not be great ways for you to do that (I do it by manually looking at the MySQL table).  One way you can get a sense of what's being stored though is to use the new field in an event/article, then move on to a new event or article, then return to the first one (maybe with a browser refresh in between): you should see the info you entered the first time.

So, let me know if you notice anything fishy with this.  When you guys are happy with it I'll close the issue.

**Comment by davidskalinder** [comment]
@matthewsmorgan sez:

>     * Question: What is the purpose of "Other Location" under "City/Town"? Maybe this was put here before we remembered that Location should be on the text select list? If that's the case, I think this field is redundant and can be removed given that info such as city hall can be added under the text select section

My suggestion upthread was:

> another text box for "Other location" (to get things like "Adams St. Bridge" but also, e.g., states that for some reason don't fit into the list above even though my hope is that it's logically exhaustive).

I think I made that suggestion based on what was in the methods paper.  I suppose the worry is that that there were probably events at 20 sites in Cincinnati on that one day and we want pass 2 coders to be able to distinguish them.  But perhaps the text-select field, plus an uncertainty field on the basic info tab, will be sufficient?

@olderwoman, your opinion is probably the best-informed as far as this goes?

**Comment by olderwoman** [comment]
As a coder, when you are trying to parse events, it does help to pay attention to exactly what was where but I was doing that in the event description field. I suggest that “other location” is best left off as a field that has to be filled in and instead is a marked-text option, where coders mark all location information as part of one location field. Does that seem easy? My fear about creating an other location box is that people will feel compelled to fill it out and I am not sure it will be all that helpful, better as marked text I think.

**Comment by davidskalinder** [comment]
> My fear about creating an other location box is that people will feel compelled to fill it out and I am not sure it will be all that helpful, better as marked text I think.

Yep, sounds right to me.  I'll cut it.

**Comment by davidskalinder** [comment]
Okay, this should now be changed in the testing deployment (and in branch split_location).

**Comment by matthewsmorgan** [comment]
Yes, this works.
